source/common/ssl/context_impl.cc

```diff
@@ -370,6 +383,133 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
                                },
                                this);
   }
+
+  if (!config.sessionTicketKeys().empty()) {
+    session_ticket_keys_.resize(config.sessionTicketKeys().size());
+    for (unsigned i = 0; i < config.sessionTicketKeys().size(); ++i) {
+      // If this changes, need to figure out how to deal with key files
+      // that previously worked.  For now, just assert so we'll notice that
+      // it changed if it does.
+      static_assert(sizeof(SessionTicketKey) == 80, "Input is expected to be this size");
+
+      const std::vector<uint8_t>& src_key = config.sessionTicketKeys().at(i);
+      SessionTicketKey& dst_key = session_ticket_keys_.at(i);
+
+      if (src_key.size() != sizeof(SessionTicketKey)) {
+        throw EnvoyException(fmt::format("Incorrect TLS session ticket key length.  "
+                                         "Index {}, length {}, expected length {}.",
+                                         i, src_key.size(), sizeof(SessionTicketKey)));
+      }
+
+      std::copy_n(src_key.begin(), dst_key.name.size(), dst_key.name.begin());
+      size_t pos = dst_key.name.size();
+      std::copy_n(src_key.begin() + pos, dst_key.hmac_key.size(), dst_key.hmac_key.begin());
+      pos += dst_key.hmac_key.size();
+      std::copy_n(src_key.begin() + pos, dst_key.aes_key.size(), dst_key.aes_key.begin());
+      pos += dst_key.aes_key.size();
+      ASSERT(src_key.begin() + pos == src_key.end());
+    }
+    SSL_CTX_set_tlsext_ticket_key_cb(
+        ctx_.get(),
+        [](SSL* ssl, uint8_t* key_name, uint8_t* iv, EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx,
+           int encrypt) -> int {
+          ContextImpl* context_impl = static_cast<ContextImpl*>(
+              SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), sslContextIndex()));
+          return dynamic_cast<ServerContextImpl*>(context_impl)
+              ->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
+        });
+  }
+
+  // Hash all the settings that affect whether the server will allow/accept
+  // the client connection.  This ensures that the client is always validated against
+  // the correct settings, even if session resumption across different listeners
+  // is enabled.
+  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
+  unsigned session_context_len = 0;
+  EVP_MD_CTX md;
+  int rc = EVP_DigestInit(&md, EVP_sha256());
+  RELEASE_ASSERT(rc == 1);
+
+  // Include SERVER_SESSION_ID_CONTEXT so that if all the other verify-settings are unset
+  // we have a deterministic value.
+  rc = EVP_DigestUpdate(&md, &SERVER_SESSION_ID_CONTEXT, sizeof(SERVER_SESSION_ID_CONTEXT));
```

Comment:
Perhaps we could use `"envoy"` here? `SERVER_SESSION_ID_CONTEXT` doesn't seem to provide any value.

Also, the deterministic fallback could be used only when `ca_cert_ == nullptr`.

Reply:
I wasn't sure what the reason for SERVER_SESSION_ID_CONTEXT was; I agree it doesn't add much/any value.  I'll switch to "envoy", unless @mattklein123 has a reason it needs to remain.